library/core/class.session.php

```diff
@@ -73,28 +84,92 @@ public function addPermissions($perms) {
+            return true;
+        }
+
+        // Otherwise, check the current user's permissions against the user they want to ban/warn/etc.
+        // Throw a permission error if the session user doesn't outrank the user to be banned. Otherwise, return true.
+
+        if ($this->checkUserRankedPermission($sessionPermission, $userID) || !$this->checkRankedPermission($sessionPermission)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * Checks a user's permission level against a ranked list of permissions and implicitly assigns
+     * the user any permissions below their level even in if the permission isn't explicitly assigned.
+     * Based on checkRakedPermission().
+     *
+     * @param string $permission
+     * @param int|string $userID
+     * @return bool
+     * @throws ContainerException Throws exception if there's a problem getting the container.
+     * @throws NotFoundException Throws exception if there's a problem getting the container.
+     */
+    public function checkUserRankedPermission(string $permission, $userID) {
+        $rankedPermissions = self::RANKED_PERMISSIONS;
+
+        $userModel = Gdn::getContainer()->get(UserModel::class);
+        $userPermissions = $userModel->getPermissions($userID);
+        $userPermissions = $userPermissions->getPermissions();
+        // Get the highest permission in the user's permission array.
+
+        if ($userPermissions[$permission] === true) {
+            return true;
+        } elseif ($userPermissions[$permission] === false) {
+            return false;
+        } elseif (in_array($permission, $rankedPermissions)) {
+            // Ordered rank of some permissions, highest to lowest
+            $currentPermissionRank = array_search($permission, $rankedPermissions);
+
+            /**
+             * If the current permission is in our ranked list, iterate through the list, starting from the highest
+             * ranked permission down to our target permission, and determine if any are applicable to the current
+             * user.  This is done so that a user with a permission like Garden.Settings.Manage can still validate
+             * permissions against a Garden.Moderation.Manage permission check, without explicitly having it
+             * assigned to their role.
+             */
+            for ($i = 0; $i <= $currentPermissionRank; $i++) {
+                if (in_array($rankedPermissions[$i], $userPermissions)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
     /**
      * Check the given permission, but also return true if the user has a higher permission.
      *
      * @param bool|string $permission The permission to check.  Bool to force true/false.
      * @return boolean True on valid authorization, false on failure to authorize
      */
     public function checkRankedPermission($permission) {
```

Comment:
I think this can be extracted to the `Permission` class so it can be used by the session or user model.